crates/starknet-devnet-core/src/starknet/mod.rs

```diff
@@ -793,6 +793,39 @@ impl Starknet {
         }
     }
 
+    pub fn update_gas(&mut self, gas_prices: GasUpdate) -> DevnetResult<GasUpdate> {
+        // BlockContext needs to be reinitialized
+        self.block_context = Starknet::init_block_context(
+            gas_prices.gas_price_wei,
+            gas_prices.gas_price_strk,
+            gas_prices.data_gas_price_wei,
+            gas_prices.data_gas_price_strk,
+            constants::ETH_ERC20_CONTRACT_ADDRESS,
+            constants::STRK_ERC20_CONTRACT_ADDRESS,
+            ChainId::from(self.block_context.chain_info().chain_id.clone()),
+            self.block_context.block_info().block_number.0,
+        );
+
+        // Pending block header gas data needs to be updated
+        self.blocks.pending_block.header.l1_gas_price.price_in_wei =
```

Comment:
if the pending block has already some transaction, what will happen with already computed fees?

Reply:
In block on demand mode you can update gas every transaction.

I was thinking about this case, we could mine block before gas update but I'm not sure what it's better for users.

Reply:
Would it make sense to apply the new gas price from the next block? And potentially the user could specify if they want the gas price change to create a new block immediately or not (similar to time setting).

Reply:
I think that changes should apply to current pending block/state and this should stay as it is but I also added generate_block flag

8af17c4f55c68f932f34aed414eb3b3c89cc6a38

Reply:
do you mean that if there are already transactions in the block the gas prices can change before block is mined

Reply:
Yes it's possible, is there something wrong with this approach?

Reply:
gas prices are set per block they cannot be tweaked before block is mined/closed

Reply:
I can propose some in-memory struct to track these gas changes and prevent double gas change in same block or just  change the gas and mine block

@FabijanC what do you think?

Reply:
I haven't yet reviewed, but the proposal from my comment in this conversation implied 1 block = 1 gas price. I meant that in any case the user would specify the gas price for the next block, and optionally it could be specified whether to generate the block right away or not.

But are we sure that 1 block = 1 gas price? Base fee should not be confused with gas price.

Reply:
you talk about ethereum or starknet?

Reply:
starknet, but I feel I'm mixing the terminology, feel free to correct me

Reply:
https://docs.starknet.io/architecture-and-concepts/network-architecture/block-structure/

Please check `l1_gas_price` and `l1_data_gas_price` from my understanding these 2 parameters are fixed in a single block

Reply:
so do you want to update gas from the next block? Should I start changing that?

Reply:
The following request gets all transactions in a block. Judging from it the gas price is set per block. If you execute multiple times the request and observe the `l1_gas_price` and `transactions` fields you will see that `l1_gas_price` field doesnt change, but `transactions` changes

curl --location 'https://free-rpc.nethermind.io/mainnet-juno/v0_6' \
--header 'Content-Type: application/json' \
--data '{
    "jsonrpc": "2.0",
    "id": "1",
    "method": "starknet_getBlockWithTxHashes",
    "params": {
        "block_id": "pending"
    }
}'